### ISSUE TITLE
epoll (and perhaps other reactors) busy loops when file descriptor limit is hit

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5368]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5368|
|Type|defect|
|Created|2011-11-05 13:08:12Z|
|Branch|https://github.com/twisted/twisted/tree/5368-break-on-emfile|

Lifted from the comments in [#11553](https://github.com/twisted/twisted/issues/11553) - note that [#11553](https://github.com/twisted/twisted/issues/11553) covers the `select()`-specific issue of `FD_SETSIZE` limits, whereas this is related to running out of file descriptors when doing `accept()`:

    If you run out of FDs in twisted and you've got a connection that you're about to accept twisted will get stuck in an infinite busy loop. epoll will signal that theres data pending on a certain fd, tcp.py will attempt to open the connection but will fail with EMFILE/ENFILE ... then epoll will immediately signal on that fd again. 

Allowing a user policy to do something useful in this case (closing idle connections, say) is covered by [#2122](https://github.com/twisted/twisted/issues/2122). This ticket is focused merely on preventing the 100% CPU usage.

<details><summary>Searchable metadata</summary>

```
trac-id__5368 5368
type__defect defect
reporter__itamar itamar
priority__high high
milestone__ 
branch__5368_break_on_emfile 5368-break-on-emfile
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__None None
time__1320498492000000 1320498492000000
changetime__1524081909259275 1524081909259275
version__None None
owner__markrwilliams markrwilliams
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst commented|
|-|-|

(In [33078]) Branching to 'EMFILE-5368'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

Ready for review. Tests:

http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/EMFILE-5368

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  It'll be pretty cool when reactors are more robust in resource exhaustion conditions.

 1. That said, a new implicit state for tcp Ports doesn't sound like a step in the right direction.  What happens if application code decides to pause the port in the one second disabled interval?  This code will silently unpause it, probably surprising the application.
 1. The timed call outliving the port's listening lifetime is also potentially troublesome.  What if the port is stopped and then started again across the one second disabled interval?  Even more fun, what if it runs into EMFILE a second time before the `restore` delayed call fires?
 1. I suppose you can address these points by just being more careful about cleaning up state across state transitions.  Maybe that's even okay, since we're going to move on to get rid of this code and replace it with application hooks for deciding what to do.  I hope we do that soon.  Maybe the literal 1 hard coded in several places can be replaced by a symbolic constant of some sort, though.
 1. Thanks for cleaning up the use of the global reactor in `SelectReactorTestCase`/`ResourceExhaustionTestCase`.  The tests are still posixbase-specific though, since they directly instantiate `twisted.internet.tcp.Port`.  Maybe not now, but at some point I think the module should be renamed/relocated to reflect this.
 1. The `ResourceExhaustionTestCase.ports` list is unused now and can be removed.
 1. `ResourceExhaustionTestCase._acceptFailureTest` doesn't need to clean up the port it creates.
 1. The `shouldStopListening` argument to that method needs some documentation.
 1. Those `FakeSocket` and `FakeReactor` should probably be generalized, tested, and re-used elsewhere in our test suite.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

I don't think we're going replace this code with user hooks; the user hooks are an additional, different part of the solution.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst commented|
|-|-|

Another potential solution is to close the socket and immediately reopen it, thus (presumably) clearing any notifications. Unfortunately some other thread might grab this last available file descriptor in between the two steps, so probably not doable in practice.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

> I don't think we're going replace this code with user hooks; the user hooks are an additional, different part of the solution.

Why aren't the user hooks enough?  This feature doesn't need access to any nasty `tcp.Port` guts.  Implemented on the factory or another application-specified object, it's simpler and more general (consider UNIX sockets, or any other source of new file descriptors), as well as customizable.

> Another potential solution is to close the socket and immediately reopen it, thus (presumably) clearing any notifications.

That doesn't sound very desirable.  There are all kinds of other cases where it's not possible to re-open the socket properly (maybe we lost the privilege to bind the address it's listening too, maybe we're sharing it with another process, etc).

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

As far as user hooks, I was thinking more in terms of optional user hooks, which 99% users won't use and therefore we're back to square one in terms of vulnerability. But, if there's a default that does the same thing as this, only using the user hook, then yes, that seems better.

In which case, maybe I should write the user hook first.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst commented|
|-|-|

Once [#5400](https://github.com/twisted/twisted/issues/5400) is merged, an additional ticket should be opened for moving TCP ports to an explicit state machine. Once *that* is done, the concerns raised in the review should be much easier to address.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_11"></a>|@itamarst commented|
|-|-|

Moving to state machine is going to take a while. How about this approach? On startup, grab extra file descriptor. If we get out-of-fds error in accept(), close the extra fd, do accept() as usual, except closing the sockets we get back from accept(), and then re-grab an extra file descriptor. Doing accept() will clear the read flag on the port socket.

Obviously this privileges existing connections, but improving that is covered by that other ticket for having better, user-defined policies.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_12"></a>|@markrwilliams set owner to @exarkun|
|-|-|

https://github.com/twisted/twisted/pull/996/

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

Should have been marked for review.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @markrwilliams|
|-|-|

Reviewed.

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)|@markrwilliams set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @markrwilliams|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_17"></a>|@markrwilliams commented|
|-|-|

I'm putting this back in review even though the Codecov report is out of date; https://codecov.io/gh/twisted/twisted/pull/996/diff stops at https://github.com/twisted/twisted/pull/996/commits/9ce61b6c99a47d049af26e92a31d4efa82095b41 but https://codecov.io/gh/twisted/twisted/pull/996/commits has the branch's HEAD: https://codecov.io/gh/twisted/twisted/commit/3bfe23fffcc45edb6604ccfb00ec8e17030175c9.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @markrwilliams<br>@exarkun set status to `new`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)|@markrwilliams set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @markrwilliams|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="markrwilliams's avatar" src="https://avatars.githubusercontent.com/u/704247?s=50" width="50" height="50">](https://github.com/markrwilliams)<a name="note_22"></a>|@markrwilliams set status to `closed`|
|-|-|

In changeset f0b04e45bbbbba44c9071f5aa86cc0dc3bbd867b
```
#!CommitTicketReference repository="" revision="f0b04e45bbbbba44c9071f5aa86cc0dc3bbd867b"
Merge pull request #996 from twisted/5368-break-on-emfile

Author: markrwilliams

Reviewers: exarkun

Fixes: ticket:5368

On UNIX-like platforms, Twisted attempts to recover from EMFILE
when accepting connections on TCP and UNIX ports by shedding
incoming clients.
```
